File: 3longestSubstringWithoutRepeatingCharacters/longestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    s = [...s]
    let [result, left, right] = Array(3).fill(0)
    const characters = new Set
    while (right < s.length)
        if (!characters.has(s[right]))
        {
            characters.add(s[right++])
            result = Math.max(result, characters.size)
        }
        else characters.delete(s[left++])
    return result
};

File: 3longestSubstringWithoutRepeatingCharacters/longestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        var result = 0;
        var left = 0;
        var right = 0;
        final var characters = new HashSet<Integer>();
        while (right < s.codePointCount(0, s.length()))
            if (!characters.contains(s.codePointAt(right)))
            {
                characters.add(s.codePointAt(right++));
                result = Math.max(result, characters.size());
            }
            else characters.remove(s.codePointAt(left++));
        return result;
    }
}
